src/rpc/methods/f3.rs

```diff
@@ -0,0 +1,387 @@
+    rpc::{ApiPaths, Ctx, Permission, RpcMethod, ServerError},
+    shim::{
+        address::{Address, Protocol},
+        clock::ChainEpoch,
+    },
+};
+use fil_actor_interface::{
+    convert::{
+        from_policy_v13_to_v10, from_policy_v13_to_v11, from_policy_v13_to_v12,
+        from_policy_v13_to_v14, from_policy_v13_to_v9,
+    },
+    miner, power,
+};
+use fvm_ipld_blockstore::Blockstore;
+use num::Signed as _;
+use std::{fmt::Display, sync::Arc};
+
+pub enum GetTipsetByEpoch {}
+impl RpcMethod<1> for GetTipsetByEpoch {
+    const NAME: &'static str = "F3.GetTipsetByEpoch";
+    const PARAM_NAMES: [&'static str; 1] = ["epoch"];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (ChainEpoch,);
+    type Ok = F3TipSet;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore>,
+        (epoch,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let ts = ctx.chain_index().tipset_by_height(
+            epoch,
+            ctx.chain_store().heaviest_tipset(),
+            ResolveNullTipset::TakeOlder,
+        )?;
+        Ok(ts.into())
+    }
+}
+
+pub enum GetTipset {}
+impl RpcMethod<1> for GetTipset {
+    const NAME: &'static str = "F3.GetTipset";
+    const PARAM_NAMES: [&'static str; 1] = ["tipset_key"];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (F3TipSetKey,);
+    type Ok = F3TipSet;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore>,
+        (f3_tsk,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let tsk = f3_tsk.try_into()?;
+        let ts = ctx.chain_index().load_required_tipset(&tsk)?;
+        Ok(ts.into())
+    }
+}
+
+pub enum GetHead {}
+impl RpcMethod<0> for GetHead {
+    const NAME: &'static str = "F3.GetHead";
+    const PARAM_NAMES: [&'static str; 0] = [];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = ();
+    type Ok = F3TipSet;
+
+    async fn handle(ctx: Ctx<impl Blockstore>, _: Self::Params) -> Result<Self::Ok, ServerError> {
+        Ok(ctx.chain_store().heaviest_tipset().into())
+    }
+}
+
+pub enum GetParent {}
+impl RpcMethod<1> for GetParent {
+    const NAME: &'static str = "F3.GetParent";
+    const PARAM_NAMES: [&'static str; 1] = ["tipset_key"];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (F3TipSetKey,);
+    type Ok = F3TipSet;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore>,
+        (f3_tsk,): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let tsk = f3_tsk.try_into()?;
+        let ts = ctx.chain_index().load_required_tipset(&tsk)?;
+        let parent = ctx.chain_index().load_required_tipset(ts.parents())?;
+        Ok(parent.into())
+    }
+}
+
+pub enum GetPowerTable {}
+impl RpcMethod<1> for GetPowerTable {
```

Comment:
That's a pretty massive method. Can it be refactored?

Reply:
There's a lot of duplicate code handling different variant of power state, v12-v14 has already been simplified with a macro. v8-v11 are all different. I guess I can update `fil-actor-states` to unify the API for v8-v11 to future simplify the code with a single macro, how about doing that in a subsequent PR since it requires a release of `fil-actor-states`

Reply:
https://github.com/ChainSafe/fil-actor-states/issues/324

Reply:
Does it require a release or can we move that logic entirely to Forest's shim? In any case, I'm good with that as long as it's indeed a follow-up and not an abandonware issue. :)

Reply:
Yeah I can move the entire thing to forest as well